libcontainer/cgroups/fs2/memory.go

```diff
@@ -136,26 +136,37 @@ func getMemoryDataV2(path, name string) (cgroups.MemoryData, error) {
 	if name != "" {
 		moduleName = "memory." + name
 	}
-	usage := moduleName + ".current"
-	limit := moduleName + ".max"
-
-	value, err := fscommon.GetCgroupParamUint(path, usage)
-	if err != nil {
-		if name != "" && os.IsNotExist(err) {
-			// Ignore EEXIST as there's no swap accounting
-			// if kernel CONFIG_MEMCG_SWAP is not set or
-			// swapaccount=0 kernel boot parameter is given.
-			return cgroups.MemoryData{}, nil
+	files := []struct {
+		name  string
+		value *uint64
+	}{
+		{
+			name:  moduleName + ".current",
+			value: &memoryData.Usage,
+		},
+		{
+			name:  moduleName + ".max",
+			value: &memoryData.Limit,
+		},
+		{
+			name:  moduleName + ".peak",
+			value: &memoryData.MaxUsage,
+		},
+	}
+
+	for _, f := range files {
+		value, err := fscommon.GetCgroupParamUint(path, f.name)
+		if err != nil {
+			if name != "" && os.IsNotExist(err) {
+				// Ignore EEXIST as there's no swap accounting
+				// if kernel CONFIG_MEMCG_SWAP is not set or
+				// swapaccount=0 kernel boot parameter is given.
+				return cgroups.MemoryData{}, nil
```

Comment:
I think this line should be questioned. I think it'd be better if we continued to collect if there was an error and returned whatever we found at the end

Reply:
Not quite. This is for case of name == "swap" (which is implicit here), meaning we're trying to read `memory.swap.*` values -- and they all are either present or not, so we bail out early when getting ENOENT from open("memory.swap.current").

OTOH if `*.peak` counter is optional, the logic should be changed. Is it?

While at it, we can always return `memoryData` not `cgroups.MemoryData`.

(BTW I took a look at `CONFIG_MEMCG_SWAP` and it was dropped from Linux v6.1, but there's still `CONFIG_SWAP`)

Reply:
> Not quite. This is for case of name == "swap" (which is implicit here), meaning we're trying to read `memory.swap.*` values -- and they all are either present or not, so we bail out early when getting ENOENT from open("memory.swap.current").

Meaning, I'd drop the second commit in this PR (and maybe add a better comment explaining why we're bailing out early).

Reply:
*peak is added in kernel 5.14--so not yet in any el variants (it's being backported to centos 9 now) which is why I think the second commit is the correct behavior

Reply:
For peak, yes. So it should be like this:

- if `memory.swap.current` is not available, return empty set and no error (that's what the current code does);
- if `*.peak` is not available, return what we have collected so far.

Reply:
So, maybe something like this

```go
if err != nil {
   if os.IsNotExist(err) {
      // Swap accounting is optional, so return early if absent.
      if name == "swap" {
         err = nil
      }
      // Peak usage is only available since kernel v5.14.
      if strings.HasSuffix(v.name, "peak") {
         err = nil
      }
     return memoryData, err
   }
}
```

(plus add a comment above saying the order is important)

Reply:
Or, for better readability,
```diff
-      if name == "swap" { 
+      if name == "swap" { // memory.swap.current not found.
```
or even
```diff
-      if name == "swap" { 
+      if f.name == "memory.swap.current" {
```